Platform/OpenNtfsDxe/Disc.c

```diff
@@ -0,0 +1,731 @@
+/** @file
+  Copyright (c) 2022, Mikhail Krichanov. All rights reserved.
+  SPDX-License-Identifier: BSD-3-Clause
+
+  Functional and structural descriptions follow NTFS Documentation
+  by Richard Russon and Yuval Fledel
+**/
+
+#include "NTFS.h"
+#include "Helper.h"
+
+UINTN mFileRecordSize;
+UINTN mIndexRecordSize;
+UINTN mSectorSize;
+UINTN mClusterSize;
+
+EFI_STATUS
+NtfsDir (
+  IN  EFI_FS        *FileSystem,
+  IN  CONST CHAR16  *Path,
+  OUT EFI_NTFS_FILE *File,
+  IN  FUNCTION_TYPE FunctionType
+  )
+{
+  EFI_STATUS   Status;
+  NTFS_FILE    *Dir = NULL;
+
+  CopyMem (&File->RootFile, FileSystem->RootIndex, sizeof (NTFS_FILE));
+  CopyMem (&File->MftFile, FileSystem->MftStart, sizeof (NTFS_FILE));
+
+  Status = FsHelpFindFile (
+    Path,
+    &File->RootFile,
+    &Dir,
+    FSHELP_DIR
+    );
+  if (EFI_ERROR(Status)) {
+    return Status;
+  }
+
+  Status = IterateDir (Dir, File, FunctionType);
+  if (EFI_ERROR(Status)) {
+    return Status;
+  }
+
+  return EFI_SUCCESS;
+}
+
+EFI_STATUS
+NtfsOpen (
+  IN EFI_NTFS_FILE *File
+  )
+{
+  EFI_STATUS   Status;
+  NTFS_FILE    *BaseMftRecord = NULL;
+
+  CopyMem (&File->RootFile, File->FileSystem->RootIndex, sizeof (NTFS_FILE));
+  CopyMem (&File->MftFile, File->FileSystem->MftStart, sizeof (NTFS_FILE));
+
+  Status = FsHelpFindFile (
+    File->Path,
+    &File->RootFile,
+    &BaseMftRecord,
+    FSHELP_REG
+    );
+  if (EFI_ERROR(Status)) {
+    return Status;
+  }
+
+  if (BaseMftRecord != &File->RootFile) {
+    CopyMem (&File->RootFile, BaseMftRecord, sizeof (*BaseMftRecord));
+    FreeFile (BaseMftRecord);
+
+    if (!File->RootFile.InodeRead) {
+      Status = InitFile (&File->RootFile, File->RootFile.Inode);
+      if (EFI_ERROR(Status)) {
+        return Status;
+      }
+    }
+  }
+
+  File->Offset = 0;
+
+  return EFI_SUCCESS;
+}
+
+EFI_STATUS
+NtfsMount (
+  IN EFI_FS   *FileSystem
+  )
+{
+  EFI_STATUS         Status;
+  BOOT_FILE_DATA     Boot;
+  UINTN              Size;
+  EFI_NTFS_FILE      *RootFile;
+
+  Status = DiskRead (FileSystem, 0, sizeof (BOOT_FILE_DATA), &Boot);
+  if (EFI_ERROR(Status)) {
+    return Status;
+  }
+
+  if ((CompareMem (Boot.SystemId, "NTFS", 4) != 0)
```

Comment:
Why not `AsciiStrnCmp`?

Reply:
AsciiStrnCmp is buggy and will try to determine lengths of its arguments. Boot.SystemId may not have '\0'.
```
INTN
EFIAPI
AsciiStrnCmp (
  IN      CONST CHAR8  *FirstString,
  IN      CONST CHAR8  *SecondString,
  IN      UINTN        Length
  )
{
  if (Length == 0) {
    return 0;
  }

  //
  // ASSERT both strings are less long than PcdMaximumAsciiStringLength
  //
  ASSERT (AsciiStrSize (FirstString));
  ASSERT (AsciiStrSize (SecondString));
```

Reply:
I see, thanks. Then please change `4` to `L_STR_LEN ("NTFS")` at least.